File: packages/cli/src/commands/agents.ts
import { Command } from 'commander';
import { icons, c, log, logJson, getOutputMode } from '../lib/output.js';
import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';

// -- Types --

interface AgentInfo {
  name: string;
  skillsDir: string;
  detected: boolean;
  skillCount: number;
}

// -- Agent definitions --

const KNOWN_AGENTS: Array<{ name: string; dir: string }> = [
  { name: 'Claude Code', dir: path.join(os.homedir(), '.claude', 'skills') },
  { name: 'Cursor', dir: path.join(os.homedir(), '.cursor', 'skills') },
  { name: 'Codex', dir: path.join(os.homedir(), '.agents', 'skills') },
  { name: 'Copilot', dir: path.join(os.homedir(), '.copilot', 'skills') },
  { name: 'Gemini CLI', dir: path.join(os.homedir(), '.gemini', 'skills') },
];

// -- Helpers --

const detectAgent = (agentDef: { name: string; dir: string }): AgentInfo => {
  const exists = fs.existsSync(agentDef.dir);
  let skillCount = 0;

  if (exists) {
    try {
      const entries = fs.readdirSync(agentDef.dir, { withFileTypes: true });
      skillCount = entries.filter((e) => e.isDirectory()).length;
    } catch {
      // Directory exists but can't be read
    }
  }

  return {
    name: agentDef.name,
    skillsDir: agentDef.dir,
    detected: exists,
    skillCount,
  };
};

const shortenPath = (fullPath: string): string => {
  const home = os.homedir();
  if (fullPath.startsWith(home)) {
    return '~' + fullPath.slice(home.length);
  }
  return fullPath;
};

// -- Command --

export const registerAgentsCommand = (program: Command): void => {
  program
    .command('agents')
    .description('Detect installed AI agents and linked skills')
    .action(async () => {
      const mode = getOutputMode();
      if (mode === 'silent') return;

      const agents = KNOWN_AGENTS.map(detectAgent);

      if (mode === 'json') {
        logJson(agents);
        return;
      }

      log('');
      log('  Detected agents:');

      for (const agent of agents) {
        if (agent.detected) {
          const nameCol = c.name(agent.name.padEnd(15));
          const pathCol = c.dim(shortenPath(agent.skillsDir).padEnd(28));
          const countStr = `${agent.skillCount} skill${agent.skillCount === 1 ? '' : 's'} linked`;
          log(`    ${icons.success} ${nameCol}${pathCol}${countStr}`);
        } else {
          const nameCol = c.dim(agent.name.padEnd(15));
          log(`    ${icons.pending} ${nameCol}${c.dim('(not detected)')}`);
        }
      }

      const totalSkills = agents.reduce((sum, a) => sum + a.skillCount, 0);
      const detectedCount = agents.filter((a) => a.detected).length;

      log('');
      if (detectedCount > 0 && totalSkills > 0) {
        log(
          `  ${totalSkills} skill${totalSkills === 1 ? '' : 's'} linked across ${detectedCount} agent${detectedCount === 1 ? '' : 's'}`,
        );
      } else if (detectedCount > 0) {
        log(
          `  ${detectedCount} agent${detectedCount === 1 ? '' : 's'} detected, no skills linked yet`,
        );
        log(`  ${c.hint('Install a skill: spm install <name>')}`);
      } else {
        log(`  ${icons.info} No agents detected`);
        log(`  ${c.hint('Install an AI agent to get started.')}`);
      }
      log('');
    });
};

File: cli-go/cmd/agents.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/skillpkg/spm/internal/output"
	"github.com/spf13/cobra"
)

// agentDef defines a known AI agent and its skills directory.
type agentDef struct {
	Name string
	Dir  string
}

// agentResult holds detection results for an agent.
type agentResult struct {
	Name       string `json:"name"`
	SkillsDir  string `json:"skills_dir"`
	Detected   bool   `json:"detected"`
	SkillCount int    `json:"skill_count"`
}

// knownAgentsFunc allows overriding for tests.
var knownAgentsFunc = defaultKnownAgents

func defaultKnownAgents() []agentDef {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil
	}
	return []agentDef{
		{Name: "Claude Code", Dir: filepath.Join(home, ".claude", "skills")},
		{Name: "Cursor", Dir: filepath.Join(home, ".cursor", "skills")},
		{Name: "Codex", Dir: filepath.Join(home, ".agents", "skills")},
		{Name: "Copilot", Dir: filepath.Join(home, ".copilot", "skills")},
		{Name: "Gemini CLI", Dir: filepath.Join(home, ".gemini", "skills")},
	}
}

var agentsCmd = &cobra.Command{
	Use:   "agents",
	Short: "Detect installed AI agents and linked skills",
	Long:  "Check known agent directories for installed AI agents and count linked skills.",
	Args:  cobra.NoArgs,
	RunE:  runAgents,
}

func init() {
	rootCmd.AddCommand(agentsCmd)
}

func detectAgent(def agentDef) agentResult {
	result := agentResult{
		Name:      def.Name,
		SkillsDir: def.Dir,
	}

	info, err := os.Stat(def.Dir)
	if err != nil || !info.IsDir() {
		return result
	}

	result.Detected = true

	entries, err := os.ReadDir(def.Dir)
	if err != nil {
		return result
	}

	for _, e := range entries {
		if e.IsDir() {
			result.SkillCount++
		}
	}

	return result
}

func shortenPath(fullPath string) string {
	home, err := os.UserHomeDir()
	if err != nil {
		return fullPath
	}
	if strings.HasPrefix(fullPath, home) {
		return "~" + fullPath[len(home):]
	}
	return fullPath
}

func runAgents(_ *cobra.Command, _ []string) error {
	agents := knownAgentsFunc()
	results := make([]agentResult, 0, len(agents))
	for _, a := range agents {
		results = append(results, detectAgent(a))
	}

	if Out.Mode == output.ModeJSON {
		return Out.LogJSON(results)
	}

	Out.Log("")
	Out.Log("  Detected agents:")
	Out.Log("")

	detectedCount := 0
	totalSkills := 0

	for _, r := range results {
		if r.Detected {
			detectedCount++
			totalSkills += r.SkillCount
			nameStr := padRight(r.Name, 15)
			pathStr := padRight(shortenPath(r.SkillsDir), 30)
			countStr := fmt.Sprintf("%d skill%s linked", r.SkillCount, pluralS(r.SkillCount))
			Out.Log("    %s %s%s%s", output.Green(output.Icons["success"]), output.Cyan(nameStr), output.Dim(pathStr), countStr)
		} else {
			nameStr := padRight(r.Name, 15)
			Out.Log("    %s %s%s", output.Dim("-"), output.Dim(nameStr), output.Dim("(not detected)"))
		}
	}

	Out.Log("")
	if detectedCount > 0 && totalSkills > 0 {
		Out.Log("  %d skill%s linked across %d agent%s", totalSkills, pluralS(totalSkills), detectedCount, pluralS(detectedCount))
	} else if detectedCount > 0 {
		Out.Log("  %d agent%s detected, no skills linked yet", detectedCount, pluralS(detectedCount))
		Out.Log("  %s", output.Dim("Install a skill: spm install <name>"))
	} else {
		Out.Log("  %s No agents detected", output.Icons["info"])
		Out.Log("  %s", output.Dim("Install an AI agent to get started."))
	}
	Out.Log("")

	return nil
}
